pkg/volume/flexvolume/probe.go

```diff
@@ -19,28 +19,185 @@ package flexvolume
+	prober.testAndSetProbeNeeded(true)
+	prober.lastUpdated = time.Now()
 
-	files, _ := ioutil.ReadDir(pluginDir)
+	if err := prober.createPluginDir(); err != nil {
+		return err
+	}
+	if err := prober.initWatcher(); err != nil {
+		return err
+	}
+
+	go func() {
+		defer prober.watcher.Close()
+		for {
+			select {
+			case event := <-prober.watcher.Events:
+				if err := prober.handleWatchEvent(event); err != nil {
+					glog.Errorf("Flexvolume prober watch: %s", err)
+				}
+			case err := <-prober.watcher.Errors:
+				glog.Errorf("Received an error from watcher: %s", err)
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Probes for Flexvolume drivers.
+// If a filesystem update has occurred since the last probe, updated = true
+// and the list of probed plugins is returned.
+// Otherwise, update = false and probedPlugins = nil.
+//
+// If an error occurs, updated and plugins are set arbitrarily.
+func (prober *flexVolumeProber) Probe() (updated bool, plugins []volume.VolumePlugin, err error) {
+	probeNeeded := prober.testAndSetProbeNeeded(false)
+
+	if !probeNeeded {
+		return false, nil, nil
+	}
+
+	files, err := ioutil.ReadDir(prober.pluginDir)
+	if err != nil {
+		return false, nil, fmt.Errorf("Error reading the Flexvolume directory: %s", err)
+	}
+
+	plugins = []volume.VolumePlugin{}
+	allErrs := []error{}
 	for _, f := range files {
 		// only directories are counted as plugins
 		// and pluginDir/dirname/dirname should be an executable
 		// unless dirname contains '~' for escaping namespace
 		// e.g. dirname = vendor~cifs
 		// then, executable will be pluginDir/dirname/cifs
 		if f.IsDir() {
-			plugin, err := NewFlexVolumePlugin(pluginDir, f.Name())
-			if err != nil {
+			plugin, pluginErr := NewFlexVolumePlugin(prober.pluginDir, f.Name())
+			if pluginErr != nil {
+				pluginErr = fmt.Errorf(
+					"Error creating Flexvolume plugin from directory %s, skipping. Error: %s",
+					f.Name(), pluginErr)
+				allErrs = append(allErrs, pluginErr)
 				continue
 			}
 
 			plugins = append(plugins, plugin)
 		}
 	}
-	return plugins
+
+	return true, plugins, errors.NewAggregate(allErrs)
+}
+
+func (prober *flexVolumeProber) handleWatchEvent(event fsnotify.Event) error {
+	eventPathAbs, err := filepath.Abs(event.Name)
+	if err != nil {
+		return err
+	}
+
+	pluginDirAbs, err := filepath.Abs(prober.pluginDir)
+	if err != nil {
+		return err
+	}
+
+	if eventOpIs(event, fsnotify.Remove) && eventPathAbs == pluginDirAbs {
+		// pluginDir needs to exist in order to be watched.
+		if err := prober.createPluginDir(); err != nil {
+			return err
+		}
+		if err := prober.addWatchRecursive(pluginDirAbs); err != nil {
+			return err
+		}
+	} else if eventOpIs(event, fsnotify.Create) {
+		if err := prober.addWatchRecursive(eventPathAbs); err != nil {
+			return err
+		}
+	}
+
+	if time.Since(prober.lastUpdated) > time.Second { // Reduce the frequency of probes.
+		prober.testAndSetProbeNeeded(true)
+		prober.lastUpdated = time.Now()
+	}
+
+	return nil
```

Comment:
There is a race here if files are dropped quickly in succession, and the read happens before the 2nd event is processed.

Let's make sure to add testing to see how often this might happen, and if it is infrequent put a comment indicating you know about it.